runtime: GC: implement precise heap

aykevl opened the pull request:
This implements the block-based GC as a partially precise GC. This means that for most heap allocations it is known which words contain a pointer and which don't. This should in theory make the GC faster (because it can skip non-pointer object) and have fewer false positives in a GC cycle. It does however use a bit more RAM to store the layout of each object.

Right now this GC seems to be slower than the conservative GC, but should be less likely to run out of memory as a result of false positives.

This PR defaults to the precise GC on regular OSes (Windows/MacOS/Linux) and doesn't change the default elsewhere (baremetal, wasm, etc). This can be changed in the future if needed.

aykevl committed:
runtime: move GC code around to prepare for precise GC

Most of the code of the conservative GC can be reused for the precise
GC. So before adding precise GC support, this commit just moves code
around to make the next commit cleaner. It is a non-functional change.

aykevl committed:
runtime: arm64 actually has 16-byte alignment like amd64

Proof: https://godbolt.org/z/as4EM3713
Essentially, this means that there are objects on arm64 that have a
16-byte alignment and so we have to respect that when we allocate things
on the heap.

aykevl committed:
runtime: implement precise GC

This implements the block-based GC as a partially precise GC. This means
that for most heap allocations it is known which words contain a pointer
and which don't. This should in theory make the GC faster (because it
can skip non-pointer object) and have fewer false positives in a GC
cycle. It does however use a bit more RAM to store the layout of each
object.

Right now this GC seems to be slower than the conservative GC, but
should be less likely to run out of memory as a result of false
positives.

aykevl committed:
runtime: zero freed memory

This helps to find bugs in the GC. It does have a performance impact so
it's only enabled when asserts are enabled.

aykevl committed:
all: store data layout as little endian value

This makes it much easier to read the value at runtime, as pointer
indices are naturally little endian. It should not affect anything else
in the program.

aykevl commented:
Looks like I need to update a few tests to get this to pass CI.

dgryski commented:
Exciting! I'll give this a review as soon as I can.

aykevl commented:
Rebased to fix the merge conflicts and to fix the interp test failure.

aykevl commented on src/runtime/gc_blocks.go:
```diff
@@ -168,6 +168,9 @@ func (b gcBlock) markFree() {
 		runtimePanic("gc: markFree() was not successful")
 	}
+	if gcAsserts {
+		*(*[4]uintptr)(unsafe.Pointer(b.address())) = [4]uintptr{}
```
Oops, should have used `wordsPerBlock` instead of hardcoding the number 4.
UPDATE: fixed :)

aykevl commented:
Did a quick check to see whether other architectures are affected, and they (surprisingy) are: baremetal systems tend to have a decrease of around 16 bytes binary size and wasm tends to have an increase in binary size of 11 bytes. Probably as a result of some changes around inlining heuristics. Nothing to worry about, honestly.

Of course, where the precise GC is enabled there is a much bigger increase of 400+ bytes. This is expected, and should not be a problem at all for these systems. It might be possible to reduce this somewhat in the future so that precise GC can also be used on AVR (where false positives are especially common).

aykevl commented:
Currently investigating what appears to be a memory corruption on arm.

aykevl committed:
runtime: arm actually has 8-byte alignment

Specification:
https://developer.arm.com/documentation/dui0472/k/C-and-C---Implementation-Details/Basic-data-types-in-ARM-C-and-C--
There are multiple types that have an 8-byte alignment (long long,
double) so we need to use the same maximum alignment in TinyGo.

Fixing this is necessary for the precise GC.

aykevl commented:
Should be fixed now. The issue was that the `runtime.align` function did not correctly align objects: they should be aligned to an 8 byte boundary and not to a 4 byte boundary.

The actual bug is quite interesting. For some reason, the ARM backend used an 'or' instruction on the pointer to add 4. This is unintuitive, but is in fact correct _as long as the pointer is sufficiently aligned_. In this case it wasn't, leading to a wrong pointer being calculated.
Not sure why they use 'or' instead of 'add', I suppose it's faster on some chips or uses less energy or something.

dgryski commented on src/internal/task/task_stack.go:
Any other places in the runtime tree we should switch to using `unsafe.Add` ?

aykevl replied:
Yes, lots of places. I have a local branch that I should finish and submit...

aykevl replied:
I'll submit it after this branch (and perhaps #2640) have been merged, to avoid conflicts.

dgryski commented on src/runtime/arch_arm.go:
```diff
@@ -9,9 +9,9 @@ const TargetBits = 32
+// Align on the maximum alignment for this platform (double).
 func align(ptr uintptr) uintptr {
-	return (ptr + 3) &^ 3
+	return (ptr + 7) &^ 7
```
Did you check all the other `align` implementations to make sure they're correct?

aykevl replied:
I should have. I've updated the PR and I think all `align` functions are now correct.

dgryski commented on src/runtime/gc_precise.go:
```diff
@@ -0,0 +1,97 @@
+			runtimePanic("unknown pointer size")
+		}
+
+		// Extract fields.
```
An example of what's happening here (vis a vis splitting the layout bits) might help here.

dgryski commented on src/runtime/gc_precise.go:
```diff
@@ -0,0 +1,97 @@
+
+	// Might be a pointer. Now look at the object layout to know for sure.
+	if scanner.bitmapAddr != nil {
+		if (*(*uint8)(unsafe.Add(scanner.bitmapAddr, index/8))>>(index%8))&1 == 0 {
```
Seems like a helper function might make this more readable, especially if we could change the type from `unsafe.Pointer` to `memoryBitmap` or something.

dgryski commented on src/runtime/gc_precise.go:
```diff
@@ -0,0 +1,97 @@
+	layout := *(*uintptr)(unsafe.Pointer(block.address()))
+	if layout == 0 {
+		// Unknown layout. Assume all words in the object could be pointers.
+		scanner.size = 1
```
These seem like magic values we might want constants for?  I guess the rest of the pointer scanning logic just "works" with these both set to `1`.

aykevl replied:
I have put a big comment explaining all of this at the top that should hopefully explain what happens here.

dgryski commented on src/runtime/gc_precise.go:
```diff
@@ -0,0 +1,97 @@
+func (scanner *gcObjectScanner) nextIsPointer(word, parent, addrOfWord uintptr) bool {
+	index := scanner.index
+	scanner.index++
+	if scanner.index == scanner.size {
```
I know this is so that the `size = 1` stuff above works, but it seems to me we might like a different way handle this so that we can assert if 'nextIsPointer` is in fact called too many times and overflows the bitmap.

aykevl replied:
This is by design to support arrays. For example, if you do `make([]*byte, 10)` you'll create a datalayout of one word with size == 1 (on most architectures). The `size = 1` above merely uses this fact, it basically treats the object like it would `make([]*T, ...)` (that is, everything could be a pointer).

dgryski replied:
So reading the comment about the `size`, I'm wondering if we know that for slices of structs the padding will always work out such that we'll never get "out-of-sync" with the layout bitmap.

aykevl replied:
Well, it should. If it doesn't, that means there is a bug somewhere (which wouldn't be very surprising to me, this code hasn't been tested a lot). If you're interested, the relevant functions are `createObjectLayout` and `getPointerBitmap`.

dgryski replied:
SGTM

dgryski commented on compiler/llvm.go:
```diff
@@ -166,6 +166,7 @@ func (c *compilerContext) createObjectLayout(t llvm.Type, pos token.Pos) llvm.Va
 	// Create the global initializer.
 	bitmapBytes := make([]byte, int(objectSizeWords+7)/8)
 	bitmap.FillBytes(bitmapBytes)
+	reverseBytes(bitmapBytes) // big-endian to little-endian
```
Now that we're actually using the layout in the allocator, do we have any tests that it's correct?

aykevl replied:
There are some tests in compiler/testdata/gc.go, or do you mean something else?

dgryski commented on src/runtime/gc_precise.go:
```diff
@@ -0,0 +1,97 @@
+	return scanner.bitmap == 0
+}
+
+func (scanner *gcObjectScanner) nextIsPointer(word, parent, addrOfWord uintptr) bool {
```
I wonder if this ends up being cleaner if we split this into `next()` call that advanced the index and `IsPointer` that does the check?

dgryski commented on src/runtime/gc_blocks.go:
```diff
@@ -0,0 +1,689 @@
+			// Load the word.
+			word := *(*uintptr)(unsafe.Pointer(addr))
+
+			if !scanner.nextIsPointer(word, root.address(), addr) {
```
I feel like the fact we have the scanner and the word as two separate things here means they might get out of sync.  For example, `nextIsPointer`, but you're passing in `word`, so it should be `*this*IsPointer`, except that you need to increment the internal scanner index .... We should figure out the best API here.

aykevl replied:
Yeah I was struggling with the API. Not too happy with what I have here but I couldn't figure out something better that is also generic. Suggestions for a better API are welcome :)

dgryski reviewed (changes requested):
Pretty confident this is correct (it passes the corpus, which is always encouraging).  Just a bunch of questions.

aykevl committed:
riscv: use 16-byte alignment everywhere

Source: https://riscv.org/wp-content/uploads/2015/01/riscv-calling.pdf

aykevl committed:
xtensa: use 8-byte alignment

This is probably not necessary on Espressif chips, but let's strictly
follow the ABI to be sure.

aykevl reviewed (commented):
I've put a big comment at the top of gc_precise.go that should answer some questions you have. Please take another look.

dgryski reviewed (approved):
One more question, but otherwise LGTM.
